File: frontend/lib/services/queue_transfer_service.dart
import 'package:dio/dio.dart';
import '../config/api_config.dart';
import '../models/queue_transfer_models.dart';

/// Service for handling queue transfer operations
class QueueTransferService {
  final Dio _dio;

  QueueTransferService({Dio? dio}) : _dio = dio ?? Dio();

  /// Creates QueueTransferService with default dependencies
  static Future<QueueTransferService> create() async {
    return QueueTransferService();
  }

  /// Transfer a queue entry to another salon, service, or time slot
  Future<QueueTransferResponse> transferQueueEntry(QueueTransferRequest request) async {
    try {
      final response = await _dio.post(
        ApiConfig.getUrl('${ApiConfig.queueTransferEndpoint}/transfer'),
        data: request.toJson(),
        options: Options(headers: ApiConfig.defaultHeaders),
      );
      
      return QueueTransferResponse.fromJson(response.data);
    } on DioException catch (e) {
      if (e.response?.statusCode == 400) {
        final errorData = e.response?.data;
        if (errorData is Map<String, dynamic>) {
          return QueueTransferResponse.fromJson(errorData);
        }
      }
      throw Exception('Failed to transfer queue entry: ${e.message}');
    } catch (e) {
      throw Exception('Failed to transfer queue entry: $e');
    }
  }

  /// Get transfer suggestions for a queue entry
  Future<TransferSuggestionsResponse> getTransferSuggestions(TransferSuggestionsRequest request) async {
    try {
      final response = await _dio.post(
        ApiConfig.getUrl('${ApiConfig.queueTransferEndpoint}/suggestions'),
        data: request.toJson(),
        options: Options(headers: ApiConfig.defaultHeaders),
      );
      
      return TransferSuggestionsResponse.fromJson(response.data);
    } on DioException catch (e) {
      if (e.response?.statusCode == 400) {
        final errorData = e.response?.data;
        if (errorData is Map<String, dynamic>) {
          return TransferSuggestionsResponse.fromJson(errorData);
        }
      }
      throw Exception('Failed to get transfer suggestions: ${e.message}');
    } catch (e) {
      throw Exception('Failed to get transfer suggestions: $e');
    }
  }

  /// Check if a queue entry is eligible for transfer
  Future<TransferEligibilityResponse> checkTransferEligibility(TransferEligibilityRequest request) async {
    try {
      final response = await _dio.post(
        ApiConfig.getUrl('${ApiConfig.queueTransferEndpoint}/eligibility'),
        data: request.toJson(),
        options: Options(headers: ApiConfig.defaultHeaders),
      );
      
      return TransferEligibilityResponse.fromJson(response.data);
    } catch (e) {
      throw Exception('Failed to check transfer eligibility: $e');
    }
  }

  /// Get transfer history for a customer
  Future<List<QueueTransferAnalytics>> getCustomerTransferHistory(String customerId) async {
    try {
      final response = await _dio.get(
        ApiConfig.getUrl('${ApiConfig.queueTransferEndpoint}/history/$customerId'),
        options: Options(headers: ApiConfig.defaultHeaders),
      );
      
      return (response.data as List)
          .map((item) => QueueTransferAnalytics.fromJson(item))
          .toList();
    } catch (e) {
      throw Exception('Failed to get transfer history: $e');
    }
  }

  /// Cancel a pending transfer
  Future<bool> cancelTransfer(String transferId) async {
    try {
      final response = await _dio.post(
        ApiConfig.getUrl('${ApiConfig.queueTransferEndpoint}/cancel/$transferId'),
        options: Options(headers: ApiConfig.defaultHeaders),
      );
      
      return response.data['success'] ?? false;
    } catch (e) {
      throw Exception('Failed to cancel transfer: $e');
    }
  }

  /// Get nearby salons for transfer suggestions
  Future<List<Map<String, dynamic>>> getNearbySalonsForTransfer(
    String salonId, {
    double maxDistanceKm = 5.0,
    int maxResults = 10,
  }) async {
    try {
      final response = await _dio.get(
        ApiConfig.getUrl('${ApiConfig.queueTransferEndpoint}/nearby-salons/$salonId'),
        queryParameters: {
          'maxDistanceKm': maxDistanceKm,
          'maxResults': maxResults,
        },
        options: Options(headers: ApiConfig.defaultHeaders),
      );
      
      return List<Map<String, dynamic>>.from(response.data['salons'] ?? []);
    } catch (e) {
      throw Exception('Failed to get nearby salons: $e');
    }
  }

  /// Quick transfer to a specific salon (simplified API)
  Future<QueueTransferResponse> quickTransferToSalon({
    required String queueEntryId,
    required String targetSalonId,
    String? reason,
  }) async {
    final request = QueueTransferRequest(
      queueEntryId: queueEntryId,
      transferType: 'salon',
      targetSalonId: targetSalonId,
      reason: reason,
      maintainPosition: false,
    );
    
    return transferQueueEntry(request);
  }

  /// Quick transfer to a different service (simplified API)
  Future<QueueTransferResponse> quickTransferToService({
    required String queueEntryId,
    required String targetServiceType,
    String? reason,
  }) async {
    final request = QueueTransferRequest(
      queueEntryId: queueEntryId,
      transferType: 'service',
      targetServiceType: targetServiceType,
      reason: reason,
      maintainPosition: true, // Usually maintain position for service changes
    );
    
    return transferQueueEntry(request);
  }

  /// Get smart transfer suggestions with filtering
  Future<TransferSuggestionsResponse> getSmartSuggestions({
    required String queueEntryId,
    double? maxDistanceKm = 5.0,
    int? maxWaitTimeMinutes,
    List<String>? preferredServices,
    bool prioritizeShorterWait = true,
    bool prioritizeCloserDistance = false,
  }) async {
    final request = TransferSuggestionsRequest(
      queueEntryId: queueEntryId,
      maxDistanceKm: maxDistanceKm,
      maxWaitTimeMinutes: maxWaitTimeMinutes,
      preferredServiceTypes: preferredServices,
      includeSalonTransfers: true,
      includeServiceTransfers: preferredServices != null,
      includeTimeTransfers: false, // Future feature
    );
    
    final response = await getTransferSuggestions(request);
    
    if (response.success && response.hasSuggestions) {
      // Apply client-side filtering and sorting
      var suggestions = response.validSuggestions;
      
      if (prioritizeShorterWait) {
        suggestions.sort((a, b) => a.estimatedWaitMinutes.compareTo(b.estimatedWaitMinutes));
      } else if (prioritizeCloserDistance) {
        suggestions.sort((a, b) => a.distanceKm.compareTo(b.distanceKm));
      }
      
      return TransferSuggestionsResponse(
        success: true,
        suggestions: suggestions,
        currentSalonId: response.currentSalonId,
        currentSalonName: response.currentSalonName,
        currentPosition: response.currentPosition,
        currentEstimatedWaitMinutes: response.currentEstimatedWaitMinutes,
        generatedAt: response.generatedAt,
        errors: response.errors,
      );
    }
    
    return response;
  }

  /// Check if transfer would be beneficial
  Future<bool> isTransferBeneficial({
    required String queueEntryId,
    required String targetSalonId,
    int minTimeSavings = 10, // Minimum minutes to save
    int minPositionImprovement = 2, // Minimum positions to improve
  }) async {
    try {
      final eligibility = await checkTransferEligibility(
        TransferEligibilityRequest(
          queueEntryId: queueEntryId,
          transferType: 'salon',
          targetSalonId: targetSalonId,
        ),
      );
      
      if (!eligibility.isEligible) return false;
      
      // Get current suggestions to compare
      final suggestions = await getTransferSuggestions(
        TransferSuggestionsRequest(queueEntryId: queueEntryId),
      );
      
      if (!suggestions.success) return false;
      
      // Find suggestion for this specific salon
      final targetSuggestion = suggestions.suggestions.firstWhere(
        (s) => s.targetSalonId == targetSalonId,
        orElse: () => throw Exception('Target salon not found in suggestions'),
      );
      
      return targetSuggestion.timeImprovement >= minTimeSavings ||
             targetSuggestion.positionImprovement >= minPositionImprovement;
    } catch (e) {
      return false;
    }
  }

  /// Get transfer statistics for analytics
  Future<Map<String, dynamic>> getTransferStats() async {
    // This would typically be called with user authentication
    // For now, return mock data
    return {
      'totalTransfers': 0,
      'successfulTransfers': 0,
      'averageTimeSaved': 0,
      'averagePositionsImproved': 0,
      'mostPopularTargetSalons': <String>[],
    };
  }
}

File: backend/GrandeTech.QueueHub/GrandeTech.QueueHub.API/Application/Queues/QueueTransferService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Grande.Fila.API.Application.Queues.Models;
using Grande.Fila.API.Domain.Queues;
using Grande.Fila.API.Domain.Locations;
using Grande.Fila.API.Domain.Customers;

namespace Grande.Fila.API.Application.Queues
{
    /// <summary>
    /// Service for handling queue transfer operations
    /// </summary>
    public class QueueTransferService : IQueueTransferService
    {
        private readonly IQueueRepository _queueRepository;
        private readonly ILocationRepository _locationRepository;
        private readonly ICustomerRepository _customerRepository;
        private readonly ILogger<QueueTransferService> _logger;

        public QueueTransferService(
            IQueueRepository queueRepository,
            ILocationRepository locationRepository,
            ICustomerRepository customerRepository,
            ILogger<QueueTransferService> logger)
        {
            _queueRepository = queueRepository;
            _locationRepository = locationRepository;
            _customerRepository = customerRepository;
            _logger = logger;
        }

        public async Task<QueueTransferResponse> TransferQueueEntryAsync(
            QueueTransferRequest request, 
            CancellationToken cancellationToken = default)
        {
            var response = new QueueTransferResponse
            {
                TransferredAt = DateTime.UtcNow,
                TransferType = request.TransferType
            };

            try
            {
                // Validate request
                if (!Guid.TryParse(request.QueueEntryId, out var queueEntryId))
                {
                    response.Errors.Add("Invalid queue entry ID format");
                    return response;
                }

                // Get current queue entry
                var currentEntry = await GetQueueEntryAsync(queueEntryId, cancellationToken);
                if (currentEntry == null)
                {
                    response.Errors.Add("Queue entry not found");
                    return response;
                }

                // Check if entry is eligible for transfer
                var eligibilityCheck = await CheckTransferEligibilityAsync(
                    new TransferEligibilityRequest
                    {
                        QueueEntryId = request.QueueEntryId,
                        TransferType = request.TransferType,
                        TargetSalonId = request.TargetSalonId,
                        TargetServiceType = request.TargetServiceType
                    }, cancellationToken);

                if (!eligibilityCheck.IsEligible)
                {
                    response.Errors.AddRange(eligibilityCheck.Restrictions);
                    return response;
                }

                // Perform transfer based on type
                switch (request.TransferType.ToLower())
                {
                    case "salon":
                        return await TransferToSalonAsync(request, currentEntry, response, cancellationToken);
                    case "service":
                        return await TransferToServiceAsync(request, currentEntry, response, cancellationToken);
                    case "time":
                        return await TransferToTimeAsync(request, currentEntry, response, cancellationToken);
                    default:
                        response.Errors.Add($"Unsupported transfer type: {request.TransferType}");
                        return response;
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error transferring queue entry {QueueEntryId}", request.QueueEntryId);
                response.Errors.Add("An error occurred during transfer");
                return response;
            }
        }

        public async Task<TransferSuggestionsResponse> GetTransferSuggestionsAsync(
            TransferSuggestionsRequest request, 
            CancellationToken cancellationToken = default)
        {
            var response = new TransferSuggestionsResponse
            {
                GeneratedAt = DateTime.UtcNow
            };

            try
            {
                if (!Guid.TryParse(request.QueueEntryId, out var queueEntryId))
                {
                    response.Errors.Add("Invalid queue entry ID format");
                    return response;
                }

                var currentEntry = await GetQueueEntryAsync(queueEntryId, cancellationToken);
                if (currentEntry == null)
                {
                    response.Errors.Add("Queue entry not found");
                    return response;
                }

                // Get current queue and salon info
                var currentQueue = await _queueRepository.GetByIdAsync(currentEntry.QueueId, cancellationToken);
                var currentSalon = await _locationRepository.GetByIdAsync(currentQueue.LocationId, cancellationToken);

                response.CurrentSalonId = currentSalon.Id.ToString();
                response.CurrentSalonName = currentSalon.Name;
                response.CurrentPosition = currentEntry.Position;
                response.CurrentEstimatedWaitMinutes = CalculateEstimatedWaitTime(currentQueue, currentEntry);

                // Generate salon transfer suggestions
                if (request.IncludeSalonTransfers)
                {
                    var salonSuggestions = await GenerateSalonTransferSuggestions(
                        currentEntry, currentSalon, request, cancellationToken);
                    response.Suggestions.AddRange(salonSuggestions);
                }

                // Generate service transfer suggestions
                if (request.IncludeServiceTransfers)
                {
                    var serviceSuggestions = await GenerateServiceTransferSuggestions(
                        currentEntry, currentQueue, request, cancellationToken);
                    response.Suggestions.AddRange(serviceSuggestions);
                }

                // Sort suggestions by priority and confidence
                response.Suggestions = response.Suggestions
                    .OrderByDescending(s => GetPriorityScore(s.Priority))
                    .ThenByDescending(s => s.ConfidenceScore)
                    .ThenBy(s => s.EstimatedWaitMinutes)
                    .ToList();

                response.Success = true;
                return response;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error generating transfer suggestions for {QueueEntryId}", request.QueueEntryId);
                response.Errors.Add("An error occurred while generating suggestions");
                return response;
            }
        }

        public async Task<TransferEligibilityResponse> CheckTransferEligibilityAsync(
            TransferEligibilityRequest request, 
            CancellationToken cancellationToken = default)
        {
            var response = new TransferEligibilityResponse();

            try
            {
                if (!Guid.TryParse(request.QueueEntryId, out var queueEntryId))
                {
                    response.Restrictions.Add("Invalid queue entry ID format");
                    return response;
                }

                var currentEntry = await GetQueueEntryAsync(queueEntryId, cancellationToken);
                if (currentEntry == null)
                {
                    response.Restrictions.Add("Queue entry not found");
                    return response;
                }

                // Check basic eligibility rules
                if (currentEntry.Status == QueueEntryStatus.Called)
                {
                    response.Restrictions.Add("Cannot transfer while being called");
                    return response;
                }

                            if (currentEntry.Status == QueueEntryStatus.CheckedIn)
            {
                response.Restrictions.Add("Cannot transfer while in service");
                return response;
            }

                if (currentEntry.Status == QueueEntryStatus.Completed)
                {
                    response.Restrictions.Add("Cannot transfer completed entries");
                    return response;
                }

                // Check transfer-specific eligibility
                switch (request.TransferType.ToLower())
                {
                    case "salon":
                        await CheckSalonTransferEligibility(request, currentEntry, response, cancellationToken);
                        break;
                    case "service":
                        await CheckServiceTransferEligibility(request, currentEntry, response, cancellationToken);
                        break;
                    case "time":
                        await CheckTimeTransferEligibility(request, currentEntry, response, cancellationToken);
                        break;
                    default:
                        response.Restrictions.Add($"Unsupported transfer type: {request.TransferType}");
                        return response;
                }

                response.IsEligible = response.Restrictions.Count == 0;
                return response;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error checking transfer eligibility for {QueueEntryId}", request.QueueEntryId);
                response.Restrictions.Add("An error occurred while checking eligibility");
                return response;
            }
        }

        public async Task<BulkTransferResponse> BulkTransferAsync(
            BulkTransferRequest request, 
            CancellationToken cancellationToken = default)
        {
            var response = new BulkTransferResponse
            {
                ProcessedAt = DateTime.UtcNow
            };

            try
            {
                foreach (var queueEntryId in request.QueueEntryIds)
                {
                    var transferRequest = new QueueTransferRequest
                    {
                        QueueEntryId = queueEntryId,
                        TransferType = "salon",
                        TargetSalonId = request.TargetSalonId,
                        Reason = request.Reason,
                        MaintainPosition = request.MaintainRelativePositions
                    };

                    var transferResult = await TransferQueueEntryAsync(transferRequest, cancellationToken);
                    response.TransferResults.Add(transferResult);

                    if (transferResult.Success)
                    {
                        response.SuccessfulTransfers++;
                    }
                    else
                    {
                        response.FailedTransfers++;
                        response.Errors.AddRange(transferResult.Errors);
                    }
                }

                response.Success = response.SuccessfulTransfers > 0;
                return response;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error performing bulk transfer");
                response.Errors.Add("An error occurred during bulk transfer");
                return response;
            }
        }

        // Mock implementations for remaining methods
        public async Task<QueueTransferAnalytics?> GetTransferAnalyticsAsync(
            string transferId, 
            CancellationToken cancellationToken = default)
        {
            // TODO: Implement with actual data store
            await Task.Delay(10, cancellationToken);
            return null;
        }

        public async Task<List<QueueTransferAnalytics>> GetCustomerTransferHistoryAsync(
            string customerId, 
            CancellationToken cancellationToken = default)
        {
            // TODO: Implement with actual data store
            await Task.Delay(10, cancellationToken);
            return new List<QueueTransferAnalytics>();
        }

        public async Task<bool> CancelTransferAsync(
            string transferId, 
            CancellationToken cancellationToken = default)
        {
            // TODO: Implement transfer cancellation
            await Task.Delay(10, cancellationToken);
            return true;
        }

        public async Task<Dictionary<string, object>> GetSalonTransferStatsAsync(
            string salonId, 
            DateTime? fromDate = null, 
            DateTime? toDate = null, 
            CancellationToken cancellationToken = default)
        {
            // TODO: Implement with actual analytics
            await Task.Delay(10, cancellationToken);
            return new Dictionary<string, object>
            {
                ["totalTransfers"] = 0,
                ["successfulTransfers"] = 0,
                ["averageWaitTimeImprovement"] = 0
            };
        }

        // Private helper methods
        private async Task<QueueEntry?> GetQueueEntryAsync(Guid queueEntryId, CancellationToken cancellationToken)
        {
            // TODO: Implement actual queue entry lookup
            await Task.Delay(10, cancellationToken);
            return null; // Mock implementation
        }

        private async Task<QueueTransferResponse> TransferToSalonAsync(
            QueueTransferRequest request, 
            QueueEntry currentEntry, 
            QueueTransferResponse response, 
            CancellationToken cancellationToken)
        {
            // TODO: Implement salon transfer logic
            await Task.Delay(10, cancellationToken);
            response.Success = true;
            response.NewQueueEntryId = Guid.NewGuid().ToString();
            response.NewSalonId = request.TargetSalonId;
            response.NewPosition = 1;
            response.EstimatedWaitMinutes = 15;
            return response;
        }

        private async Task<QueueTransferResponse> TransferToServiceAsync(
            QueueTransferRequest request, 
            QueueEntry currentEntry, 
            QueueTransferResponse response, 
            CancellationToken cancellationToken)
        {
            // TODO: Implement service transfer logic
            await Task.Delay(10, cancellationToken);
            response.Success = true;
            return response;
        }

        private async Task<QueueTransferResponse> TransferToTimeAsync(
            QueueTransferRequest request, 
            QueueEntry currentEntry, 
            QueueTransferResponse response, 
            CancellationToken cancellationToken)
        {
            // TODO: Implement time transfer logic
            await Task.Delay(10, cancellationToken);
            response.Success = true;
            return response;
        }

        private async Task<List<QueueTransferSuggestion>> GenerateSalonTransferSuggestions(
            QueueEntry currentEntry, 
            Location currentSalon, 
            TransferSuggestionsRequest request, 
            CancellationToken cancellationToken)
        {
            // TODO: Implement intelligent salon suggestions
            await Task.Delay(10, cancellationToken);
            return new List<QueueTransferSuggestion>
            {
                new QueueTransferSuggestion
                {
                    SuggestionId = Guid.NewGuid().ToString(),
                    Type = "salon",
                    Title = "Shorter wait at nearby salon",
                    Description = "Transfer to Barbearia Moderna - 10 minutes less wait time",
                    TargetSalonId = Guid.NewGuid().ToString(),
                    TargetSalonName = "Barbearia Moderna",
                    EstimatedWaitMinutes = 15,
                    PositionImprovement = 3,
                    TimeImprovement = 10,
                    DistanceKm = 0.8,
                    Priority = "high",
                    ConfidenceScore = 0.85,
                    ValidUntil = DateTime.UtcNow.AddMinutes(30)
                }
            };
        }

        private async Task<List<QueueTransferSuggestion>> GenerateServiceTransferSuggestions(
            QueueEntry currentEntry, 
            Queue currentQueue, 
            TransferSuggestionsRequest request, 
            CancellationToken cancellationToken)
        {
            // TODO: Implement service transfer suggestions
            await Task.Delay(10, cancellationToken);
            return new List<QueueTransferSuggestion>();
        }

        private async Task CheckSalonTransferEligibility(
            TransferEligibilityRequest request, 
            QueueEntry currentEntry, 
            TransferEligibilityResponse response, 
            CancellationToken cancellationToken)
        {
            if (string.IsNullOrEmpty(request.TargetSalonId))
            {
                response.Restrictions.Add("Target salon ID is required for salon transfers");
                return;
            }

            if (!Guid.TryParse(request.TargetSalonId, out var targetSalonId))
            {
                response.Restrictions.Add("Invalid target salon ID format");
                return;
            }

            // Check if target salon exists and is accepting customers
            var targetSalon = await _locationRepository.GetByIdAsync(targetSalonId, cancellationToken);
            if (targetSalon == null)
            {
                response.Restrictions.Add("Target salon not found");
                return;
            }

            // TODO: Add more eligibility checks
            response.EstimatedNewPosition = 2;
            response.EstimatedNewWaitMinutes = 20;
        }

        private async Task CheckServiceTransferEligibility(
            TransferEligibilityRequest request, 
            QueueEntry currentEntry, 
            TransferEligibilityResponse response, 
            CancellationToken cancellationToken)
        {
            // TODO: Implement service transfer eligibility
            await Task.Delay(10, cancellationToken);
        }

        private async Task CheckTimeTransferEligibility(
            TransferEligibilityRequest request, 
            QueueEntry currentEntry, 
            TransferEligibilityResponse response, 
            CancellationToken cancellationToken)
        {
            // TODO: Implement time transfer eligibility
            await Task.Delay(10, cancellationToken);
        }

        private int CalculateEstimatedWaitTime(Queue queue, QueueEntry entry)
        {
            // TODO: Implement actual wait time calculation
            return Math.Max(0, (entry.Position - 1) * 15); // Mock: 15 minutes per person
        }

        private int GetPriorityScore(string priority)
        {
            return priority.ToLower() switch
            {
                "high" => 3,
                "medium" => 2,
                "low" => 1,
                _ => 0
            };
        }
    }
}
